golang/cosmos/x/vlocalchain/keeper/keeper.go

```
@@ -0,0 +1,262 @@
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (k Keeper) ExecuteTx(origCtx sdk.Context, addr string, msgs []sdk.Msg) ([]interface{}, error) {
+	if err := k.authenticateTx(msgs, addr); err != nil {
+		return nil, err
+	}
+
+	resps := make([]interface{}, len(msgs))
+
+	// CacheContext returns a new context with the multi-store branched into a cached storage object
+	// writeCache is called only if all msgs succeed, performing state transitions atomically
+	cacheCtx, writeCache := origCtx.CacheContext()
+
+	for i, msg := range msgs {
+		if m, ok := msg.(HasValidateBasic); ok {
+			if err := m.ValidateBasic(); err != nil {
+				return nil, err
+			}
+		}
+
+		protoAny, err := k.executeMsg(cacheCtx, msg)
+		if err != nil {
+			return nil, err
+		}
+
+		if err = k.cdc.UnpackAny(protoAny, &resps[i]); err != nil {
+			return nil, err
+		}
+	}
+
+	// NOTE: The context returned by CacheContext() creates a new EventManager, so events must be correctly propagated back to the current context
+	origCtx.EventManager().EmitEvents(cacheCtx.EventManager().Events())
+	writeCache()
+
+	return resps, nil
+}
+
+// Attempts to get the message handler from the router and if found will then execute the message.
+// If the message execution is successful, the proto marshaled message response will be returned.
+func (k Keeper) executeMsg(ctx sdk.Context, msg sdk.Msg) (*codectypes.Any, error) {
+	handler := k.msgRouter.Handler(msg)
+	if handler == nil {
+		return nil, fmt.Errorf("invalid message route for msg %s", sdk.MsgTypeURL(msg))
+	}
+
+	res, err := handler(ctx, msg)
+	if err != nil {
+		return nil, err
+	}
+
+	// NOTE: The sdk msg handler creates a new EventManager, so events must be correctly propagated back to the current context
+	ctx.EventManager().EmitEvents(res.GetEvents())
+
+	// Each individual sdk.Result has exactly one Msg response. We aggregate here.
+	msgResponse := res.MsgResponses[0]
```

Panic if `len(res.MsgResponses) != 1` rather than silently ignoring extra responses.